### PR TITLE
Support specifying cluster name in ClickHouse connector

### LINKED ISSUE #17307: support clickhouse cluster ddl operation

### PR DESCRIPTION
**Pull request by nelsonspark**
Currently, trino connects to Clickhouse by configuring the connection-url address. If the connection-url is configured with the chproxy address, the chproxy proxy may be the clickhouse cluster (such as 3 shards 3 replicas), not just the stand-alone clickhouse.

If you create a table at this time, not only a single clickhouse server needs to execute the table creation operation, but all nodes need to execute it. Clickhouse distributed table creation syntax: create table ck_test on cluster ck_cluster

In the same way, such ddl operations also have
ALTER TABLE ck_test on cluster ck_cluster ADD COLUMN c1 INT;
CREATE TABLE ck_tmp on cluster ck_cluster AS ck_test;
and so on .

If it is compatible with the clickhouse cluster mode, the operation of trino and clickhouse ddl is as convenient as the stand-alone version.

Fixes #17307

### CODE REVIEW

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/BaseJdbcConfig.java [hunk 1]
```diff
@@ -163,4 +165,17 @@ public boolean isTableNamesCacheTtlConsistent()
     {
         return !metadataCacheTtl.equals(CACHING_DISABLED) || tableNamesCacheTtl.isEmpty();
     }
+
+    public String getClusterName()
+    {
+        return clusterName;
+    }
+
+    @Config("cluster-name")
```
Comment (ebyhr) [inline plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/BaseJdbcConfig.java:174]: Please update `clickhouse.rst`.

File: plugin/trino-clickhouse/src/main/java/io/trino/plugin/clickhouse/ClickHouseClient.java [hunk 2]
```diff
@@ -229,6 +229,15 @@ public ClickHouseClient(
                         .build());
     }
 
+    private String getClusterInfo()
+    {
+        StringBuilder clusterInfoBuilder = new StringBuilder();
+        if (!isNullOrEmpty(clusterName)) {
+            clusterInfoBuilder.append(String.format(" ON CLUSTER %s ", clusterName));
```
Comment (ebyhr) [inline plugin/trino-clickhouse/src/main/java/io/trino/plugin/clickhouse/ClickHouseClient.java:236]: No need to mixup `StringBuilder` & `String.format`.

**Review by ebyhr — COMMENTED** [review]
Please capitalize the commit title. 
https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md#format-git-commit-messages

Also, please submit CLA if you haven't sent it.

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/BaseJdbcClient.java [hunk 3]
```diff
@@ -108,7 +108,7 @@
     protected final String identifierQuote;
     protected final Set<String> jdbcTypesMappedToVarchar;
     protected final RemoteQueryModifier queryModifier;
-    private final IdentifierMapping identifierMapping;
+    protected final IdentifierMapping identifierMapping;
```
[the remark this answers is not in the document]
Comment (nelsonspark) [reply plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/BaseJdbcClient.java]: Solved .

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/BaseJdbcConfig.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (nelsonspark) [reply plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/BaseJdbcConfig.java]: Solved.

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/BaseJdbcConfig.java [hunk 4]
```diff
@@ -163,4 +165,17 @@ public boolean isTableNamesCacheTtlConsistent()
     {
         return !metadataCacheTtl.equals(CACHING_DISABLED) || tableNamesCacheTtl.isEmpty();
     }
+
+    public String getClusterName()
```
[the remark this answers is not in the document]
Comment (nelsonspark) [reply plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/BaseJdbcConfig.java]: Solved .

File: plugin/trino-clickhouse/src/main/java/io/trino/plugin/clickhouse/ClickHouseClient.java [hunk 5]
```diff
@@ -291,20 +301,25 @@ protected String createTableSql(RemoteTableName remoteTableName, List<String> co
...
         ImmutableList.Builder<String> tableOptions = ImmutableList.builder();
         Map<String, Object> tableProperties = tableMetadata.getProperties();
-        ClickHouseEngineType engine = ClickHouseTableProperties.getEngine(tableProperties);
-        tableOptions.add("ENGINE = " + engine.getEngineType());
-        if (engine == ClickHouseEngineType.MERGETREE && formatProperty(ClickHouseTableProperties.getOrderBy(tableProperties)).isEmpty()) {
-            // order_by property is required
-            throw new TrinoException(INVALID_TABLE_PROPERTY,
-                    format("The property of %s is required for table engine %s", ClickHouseTableProperties.ORDER_BY_PROPERTY, engine.getEngineType()));
+
+        String engine = "Log";
```
[the remark this answers is not in the document]
Comment (nelsonspark) [reply plugin/trino-clickhouse/src/main/java/io/trino/plugin/clickhouse/ClickHouseClient.java]: Revert , revert to use enum clickhouse type .

File: plugin/trino-clickhouse/src/main/java/io/trino/plugin/clickhouse/ClickHouseConfig.java [hunk 6]
```diff
@@ -13,16 +13,28 @@
...
+import static com.google.common.base.Strings.isNullOrEmpty;
+import static com.google.common.base.Strings.nullToEmpty;
+
 @DefunctConfig("clickhouse.legacy-driver")
 public class ClickHouseConfig
 {
     // TODO (https://github.com/trinodb/trino/issues/7102) reconsider default behavior
     private boolean mapStringAsVarchar;
 
+    private Set<String> jdbcTypesMappedToVarchar = ImmutableSet.of();
```
Comment (ebyhr) [inline plugin/trino-clickhouse/src/main/java/io/trino/plugin/clickhouse/ClickHouseConfig.java:34]: Revert unrelated changes.

File: plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestClickHouseConnectorTest.java [hunk 7]
```diff
@@ -113,8 +113,8 @@ protected QueryRunner createQueryRunner()
     @Test
     public void testSampleBySqlInjection()
     {
-        assertQueryFails("CREATE TABLE test (p1 int NOT NULL, p2 boolean NOT NULL, x VARCHAR) WITH (engine = 'MergeTree', order_by = ARRAY['p1', 'p2'], primary_key = ARRAY['p1', 'p2'], sample_by = 'p2; drop table tpch.nation')", "(?s).*Missing columns: 'p2; drop table tpch.nation.*");
-        assertUpdate("CREATE TABLE test (p1 int NOT NULL, p2 boolean NOT NULL, x VARCHAR) WITH (engine = 'MergeTree', order_by = ARRAY['p1', 'p2'], primary_key = ARRAY['p1', 'p2'], sample_by = 'p2')");
+        assertQueryFails("CREATE TABLE test (p1 int NOT NULL, p2 boolean NOT NULL, x VARCHAR) WITH (engine = 'mergetree', order_by = ARRAY['p1', 'p2'], primary_key = ARRAY['p1', 'p2'], sample_by = 'p2; drop table tpch.nation')", "(?s).*Missing columns: 'p2; drop table tpch.nation.*");
```
Comment (ebyhr) [inline plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestClickHouseConnectorTest.java:116]: Revert unrelated change.

File: testing/trino-product-tests-launcher/src/main/resources/docker/presto-product-tests/conf/environment/multinode-clickhouse/clickhouse.properties [hunk 8]
```diff
@@ -2,3 +2,4 @@ connector.name=clickhouse
 connection-url=jdbc:clickhouse://clickhouse:8124/
 connection-user=test
 connection-password=test
+cluster-name=test_cluster
```
Comment (ebyhr) [inline testing/trino-product-tests-launcher/src/main/resources/docker/presto-product-tests/conf/environment/multinode-clickhouse/clickhouse.properties:5]: Please add all affected statements to `TestClickHouse`.

File: plugin/trino-clickhouse/src/main/java/io/trino/plugin/clickhouse/ClickHouseConfig.java [hunk 9]
```diff
@@ -35,4 +47,29 @@ public ClickHouseConfig setMapStringAsVarchar(boolean mapStringAsVarchar)
...
+        this.jdbcTypesMappedToVarchar = ImmutableSet.copyOf(Splitter.on(",").omitEmptyStrings().trimResults().split(nullToEmpty(jdbcTypesMappedToVarchar)));
+        return this;
+    }
+
+    public Optional<String> getClusterName()
+    {
+        return clusterName;
+    }
+
+    @Config("cluster-name")
```
Comment (ebyhr) [inline plugin/trino-clickhouse/src/main/java/io/trino/plugin/clickhouse/ClickHouseConfig.java:68]: Change to `clickhouse.cluster-name`

File: plugin/trino-clickhouse/src/main/java/io/trino/plugin/clickhouse/ClickHouseConfig.java [hunk 10]
```diff
@@ -35,4 +47,29 @@ public ClickHouseConfig setMapStringAsVarchar(boolean mapStringAsVarchar)
...
+    }
+
+    @Config("jdbc-types-mapped-to-varchar")
+    public ClickHouseConfig setJdbcTypesMappedToVarchar(String jdbcTypesMappedToVarchar)
+    {
+        this.jdbcTypesMappedToVarchar = ImmutableSet.copyOf(Splitter.on(",").omitEmptyStrings().trimResults().split(nullToEmpty(jdbcTypesMappedToVarchar)));
+        return this;
+    }
+
+    public Optional<String> getClusterName()
```
Comment (ebyhr) [inline plugin/trino-clickhouse/src/main/java/io/trino/plugin/clickhouse/ClickHouseConfig.java:63]: Add `@NotNull` annotation.

File: plugin/trino-clickhouse/src/main/java/io/trino/plugin/clickhouse/ClickHouseClient.java [hunk 11]
```diff
@@ -193,6 +192,8 @@
     // An empty character means that the table doesn't have a comment in ClickHouse
     private static final String NO_COMMENT = "";
 
+    private final Optional<String> clusterName;
```
Comment (ebyhr) [inline plugin/trino-clickhouse/src/main/java/io/trino/plugin/clickhouse/ClickHouseClient.java:195]: Move to under `aggregateFunctionRewriter`.

File: plugin/trino-clickhouse/src/main/java/io/trino/plugin/clickhouse/ClickHouseClient.java [hunk 12]
```diff
@@ -212,6 +213,7 @@ public ClickHouseClient(
     {
         super("\"", connectionFactory, queryBuilder, config.getJdbcTypesMappedToVarchar(), identifierMapping, queryModifier, false);
         this.uuidType = typeManager.getType(new TypeSignature(StandardTypes.UUID));
+        this.clusterName = config.getClusterName();
```
Comment (ebyhr) [inline plugin/trino-clickhouse/src/main/java/io/trino/plugin/clickhouse/ClickHouseClient.java:216]: Move to above `uuidType`.

File: plugin/trino-clickhouse/src/main/java/io/trino/plugin/clickhouse/ClickHouseClient.java [hunk 13]
```diff
@@ -229,6 +231,15 @@ public ClickHouseClient(
                         .build());
     }
 
+    private String getClusterInfo()
+    {
+        StringBuilder clusterInfoBuilder = new StringBuilder();
+        if (clusterName.isPresent() && !isNullOrEmpty(clusterName.get())) {
+            clusterInfoBuilder.append(" ON CLUSTER ").append(clusterName.get()).append(" ");
+        }
+        return clusterInfoBuilder.toString();
```
Comment (ebyhr) [inline plugin/trino-clickhouse/src/main/java/io/trino/plugin/clickhouse/ClickHouseClient.java:240]: I expect the outer spaces are handled in the caller side.
```suggestion
        return clusterName.map("ON CLUSTER %s"::formatted).orElse("");
```

File: plugin/trino-clickhouse/src/main/java/io/trino/plugin/clickhouse/ClickHouseConfig.java [hunk 14]
```diff
@@ -13,16 +13,28 @@
...
+import static com.google.common.base.Strings.nullToEmpty;
+
 @DefunctConfig("clickhouse.legacy-driver")
 public class ClickHouseConfig
 {
     // TODO (https://github.com/trinodb/trino/issues/7102) reconsider default behavior
     private boolean mapStringAsVarchar;
 
+    private Set<String> jdbcTypesMappedToVarchar = ImmutableSet.of();
+
```
Comment (ebyhr) [inline plugin/trino-clickhouse/src/main/java/io/trino/plugin/clickhouse/ClickHouseConfig.java:35]: Remove redundant empty line.

File: plugin/trino-clickhouse/src/main/java/io/trino/plugin/clickhouse/ClickHouseConfig.java [hunk 15]
```diff
@@ -35,4 +47,29 @@ public ClickHouseConfig setMapStringAsVarchar(boolean mapStringAsVarchar)
...
+    public Optional<String> getClusterName()
+    {
+        return clusterName;
+    }
+
+    @Config("cluster-name")
+    @ConfigDescription("ClickHouse cluster name")
+    public ClickHouseConfig setClusterName(String clusterName)
+    {
+        this.clusterName = isNullOrEmpty(clusterName) ? Optional.empty() : Optional.of(clusterName);
```
Comment (ebyhr) [inline plugin/trino-clickhouse/src/main/java/io/trino/plugin/clickhouse/ClickHouseConfig.java:72]:
```suggestion
        this.clusterName = Optional.ofNullable(clusterName);
```

File: plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestClickHouseConfig.java [hunk 16]
```diff
@@ -42,18 +42,24 @@
...
+                .setClusterName(""));
     }
 
     @Test
     public void testExplicitPropertyMappings()
     {
         Map<String, String> properties = ImmutableMap.<String, String>builder()
                 .put("clickhouse.map-string-as-varchar", "true")
+                .put("jdbc-types-mapped-to-varchar", "REAL")
+                .put("cluster-name", "A6ch")
```
Comment (ebyhr) [inline plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestClickHouseConfig.java:56]: nit: `A6ch` sounds the random string. I would recommend changing to `example-cluster` or something.

File: plugin/trino-clickhouse/src/main/java/io/trino/plugin/clickhouse/ClickHouseConfig.java (hunk 6 above)
Comment (nelsonspark) [reply plugin/trino-clickhouse/src/main/java/io/trino/plugin/clickhouse/ClickHouseConfig.java:34]: ClickhouseClient construct needs the action of ClickHouseConfig.getJdbcTypesMappedToVarchar , while ClickHouseConfig  don't extend BaseJdbcConfig

Comment (ebyhr) [reply plugin/trino-clickhouse/src/main/java/io/trino/plugin/clickhouse/ClickHouseConfig.java:34]: You can inject both ClickHouseConfig & BaseJdbcConfig.

**Comment by quannh-uet** [comment]
Hi @nelsonspark 
Is there any update on this PR?

**Comment by mosabua** [comment]
:wave: @nelsonspark  - this PR has become inactive. We hope you are still interested in working on it. If so, please rebase, resolve conflict and squash commits as needed.

Please let us know, and we can try to get reviewers to help with that.

We're working on closing out old and inactive PRs, so if you're too busy or this has too many merge conflicts to be worth picking back up, we'll be making another pass to close it out in a few weeks.

**Comment by vixns** [comment]
@bitsondatadev @colebow @mosabua @nelsonspark 

I'd like to see this PR merged, will you reopen this one or should I duplicate it ?
